staging/src/k8s.io/apiextensions-apiserver/test/integration/ratcheting_test.go

```diff
@@ -0,0 +1,1576 @@
+								"name":  "kube-apiserver",
+								"port":  6443,
+								"field": "value",
+							},
+							map[string]interface{}{
+								"name":  "nginx",
+								"port":  443,
+								"field": "value",
+							},
+							map[string]interface{}{
+								"name":  "etcd",
+								"port":  2379,
+								"field": "value",
+							},
+						},
+					}},
+				expectError{
+					applyPatchOperation{
+						"attempt to change one of the fields of the items which exceed maxItems",
+						myCRDV1Beta1, myCRDInstanceName, map[string]interface{}{
+							"field": []interface{}{
+								map[string]interface{}{
+									"name":  "kube-apiserver",
+									"port":  6443,
+									"field": "value",
+								},
+								map[string]interface{}{
+									"name":  "nginx",
+									"port":  443,
+									"field": "value",
+								},
+								map[string]interface{}{
+									"name":  "etcd",
+									"port":  2379,
+									"field": "value",
+								},
+								map[string]interface{}{
+									"name":  "dev",
+									"port":  8080,
+									"field": "value",
+								},
+							},
+						}}},
+				patchMyCRDV1Beta1Schema{
+					"Require even numbered port in key, remove maxItems requirement",
+					map[string]interface{}{
+						"properties": map[string]interface{}{
+							"field": map[string]interface{}{
+								"maxItems": nil,
+								"items": map[string]interface{}{
+									"properties": map[string]interface{}{
+										"port": map[string]interface{}{
+											"multipleOf": 2,
+										},
+									},
+								},
+							},
+						},
+					}},
+
+				applyPatchOperation{
+					"reorder fields without changing anything",
+					myCRDV1Beta1, myCRDInstanceName, map[string]interface{}{
+						"field": []interface{}{
+							map[string]interface{}{
+								"name":  "nginx",
+								"port":  443,
+								"field": "value",
+							},
+							map[string]interface{}{
+								"name":  "etcd",
+								"port":  2379,
+								"field": "value",
+							},
+							map[string]interface{}{
+								"name":  "kube-apiserver",
+								"port":  6443,
+								"field": "value",
+							},
+						},
+					}},
+
+				applyPatchOperation{
+					`use "invalid" keys despite changing order and changing sibling fields to the key`,
+					myCRDV1Beta1, myCRDInstanceName, map[string]interface{}{
+						"field": []interface{}{
+							map[string]interface{}{
+								"name":  "nginx",
+								"port":  443,
+								"field": "value",
+							},
+							map[string]interface{}{
+								"name":  "etcd",
+								"port":  2379,
+								"field": "value",
+							},
+							map[string]interface{}{
+								"name":  "kube-apiserver",
+								"port":  6443,
+								"field": "this is a changed value for an an invalid but grandfathered key",
+							},
+							map[string]interface{}{
+								"name":  "dev",
+								"port":  8080,
+								"field": "value",
+							},
+						},
+					}},
+			},
+		},
+		{
+			Name: "ArrayItems correlate by index",
```

Comment:
I think that behavior is generally wrong. We've never ever treated array items as correlatable by index. Atomic lists should probably be treated as atomic list IMO.

Reply:
/hold

wasn't sure if this was still outstanding

Reply:
Agreed offline to this ability needs further discussion in the KEP-there may be merits to allowing correlation by-index but for consistency and simplicity we may choose to remove this ability in the next alpha or beta and add discussion to KEP. For this first alpha release the current rule is OK.

Leaving lgtm and remove-hold to @apelisse